addons/account/tests/test_payment.py

```diff
@@ -592,3 +595,99 @@ def test_partial_payment_inv_foreign_payment_domestic(self):
         (inv1_receivable + pay_receivable).reconcile()
         self.assertEquals(inv1.amount_residual, 500)
         self.assertEquals(inv1.amount_residual_signed, 1500)
+
+    def test_payment_onchange_partner_with_multi_company_bank_accounts(self):
+        """
+            When registering a payment, the partner bank account shouldn't be from an other company.
+            The result should be that a bank account from company B is :
+                - not used for a payment in company A
+                - is used for a payment in company B
+        """
+
+        def trigger_onchange_partner(payment):
+            payment_form = Form(payment)
+            payment_form.partner_id = payment_form.partner_id
+            payment_form.save()
+
+        partner = self.env['res.partner'].create({'name': 'Test Partner'})
+        company_a = self.env.ref('base.main_company')
+        partner_b = self.env['res.partner'].create({'name': 'Test Partner B'})
+        bank_journal_a = self.env['account.journal'].create({
+            'name': 'Journal Company A',
+            'type': 'bank',
+            'code': 'a',
+            'company_id': company_a.id
+        })
+        payment_a = self.env['account.payment'].create({
+            'payment_date': time.strftime('%Y') + '-01-01',
+            'payment_method_id': self.payment_method_manual_in.id,
+            'payment_type': 'inbound',
+            'partner_type': 'customer',
+            'amount': 1000,
+            'journal_id': bank_journal_a.id,
+            'partner_id': partner.id,
+        })
+        company_b = self.env['res.company'].create({'name': 'Company B'})
+        bank_journal_b = self.env['account.journal'].create({
+            'name': 'Journal Company B',
+            'type': 'bank',
+            'code': 'b',
+            'company_id': company_b.id
+        })
+        bank_account_b = self.env["res.partner.bank"].create({
+            "acc_number": "	BE68539007547034",
+            "bank_id": self.bank.id,
+            "partner_id": partner.id,
+            "company_id": company_b.id,
+        })
+        payment_b = self.env['account.payment'].create({
+            'payment_date': time.strftime('%Y') + '-01-01',
+            'payment_method_id': self.payment_method_manual_in.id,
+            'payment_type': 'inbound',
+            'partner_type': 'customer',
+            'amount': 1000,
+            'journal_id': bank_journal_b.id,
+            'partner_id': partner.id,
+        })
+
+        # Case 1 : partner has a commercial partner with a bank account from company B
+        bank_account_b.write({
+            'partner_id': partner_b.id
+        })
+        partner.write({
+            'commercial_partner_id': partner_b.id
+        })
```

Comment:
So this is basically part of the set up? I never changes during various cases.

Reply:
Yes it is part of the set up so the payment can use self.partner_id.commercial_partner_id.bank_ids.